lib/Sema/CSGen.cpp

```diff
@@ -3292,7 +3292,7 @@ bool swift::typeCheckUnresolvedExpr(DeclContext &DC,
   ConstraintSystemOptions Options = ConstraintSystemFlags::AllowFixes;
   auto *TC = static_cast<TypeChecker*>(DC.getASTContext().getLazyResolver());
   ConstraintSystem CS(*TC, &DC, Options);
-  CleanupIllFormedExpressionRAII cleanup(TC->Context, Parent);
+  ExprCleaner cleanup(Parent);
```

Comment:
Note that ExprCleaner saves the Exprs in the expression tree up front, whereas CleanupIllFormedExpressionRAII walks the expression tree after the scope completes.

Imagine if we find a solution, but applying the solution fails. Applying the solution mutates existing expression nodes by changing their children in addition to creating new expressions, so your logic will not clean type variables out of those nodes.

It looks like a bunch of crashers failed with this patch; this might be the reason.

I think ultimately we should have applying the solution never fail, but perhaps we should also ensure that CSApply never writes type variable types into the expression either.